resources/net.go

```diff
@@ -0,0 +1,849 @@
+	Addrs   []string `yaml:"addrs"`   // list of addresses in cidr format
+	Gateway string   `yaml:"gateway"` // gateway address
+
+	iface        *iface // a struct containing the net.Interface and netlink.Link
+	unitFilePath string // the interface unit file path
+	// XXX: replace TempDir with VarDir
+	tempDir string // temporary directory for storing the pipe socket file
+}
+
+// nlChanStruct defines the channel used to send netlink messages and errors
+// to the event processing loop in Watch.
+type nlChanStruct struct {
+	msg []syscall.NetlinkMessage
+	err error
+}
+
+// iface wraps net.Interface to add additional methods.
+type iface struct {
+	*net.Interface
+	link netlink.Link
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *NetRes) Default() Res {
+	return &NetRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *NetRes) Validate() error {
+	// validate state
+	if obj.State != ifaceUp && obj.State != ifaceDown && obj.State != "" {
+		return fmt.Errorf("state must be up down or empty")
+	}
+
+	// validate network address input
+	if (obj.Addrs == nil) != (obj.Gateway == "") {
+		return fmt.Errorf("addrs and gateway must both be set or both be empty")
+	}
+	if obj.Addrs != nil {
+		for _, addr := range obj.Addrs {
+			if _, _, err := net.ParseCIDR(addr); err != nil {
+				return errwrap.Wrapf(err, "error parsing address: %s", addr)
+			}
+		}
+	}
+	if obj.Gateway != "" {
+		if g := net.ParseIP(obj.Gateway); g == nil {
+			return fmt.Errorf("error parsing gateway: %s", obj.Gateway)
+		}
+	}
+
+	// validate the interface name
+	_, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+
+	return obj.BaseRes.Validate()
+}
+
+// Init runs some startup code for this resource.
+func (obj *NetRes) Init() error {
+	var err error
+	// tmp directory for pipe socket
+	// XXX: Replace with obj.VarDir
+	if obj.tempDir, err = ioutil.TempDir("", "pipe"); err != nil {
+		return errwrap.Wrapf(err, "could not get TempDir")
+	}
+
+	// store the network interface in the struct
+	obj.iface = &iface{}
+	if obj.iface.Interface, err = net.InterfaceByName(obj.GetName()); err != nil {
+		return errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+	// store the netlink link to use as interface input in netlink functions
+	if obj.iface.link, err = netlink.LinkByName(obj.GetName()); err != nil {
+		return errwrap.Wrapf(err, "error finding link: %s", obj.GetName())
+	}
+
+	// build the path to the networkd configuration file
+	obj.unitFilePath = networkdUnitFileDir + IfacePrefix + obj.GetName() + networkdUnitFileExt
+
+	return obj.BaseRes.Init()
+}
+
+// Watch listens for events from the specified interface via a netlink socket.
+// XXX: currently gets events from ALL interfaces, would be nice to reject
```

Comment:
XXX should mean "this is bad, fix ASAP". Is it that bad?? If not maybe a FIXME (good thing we should do) or a TODO (nice to have one day)